File: k9mail-library/src/main/java/com/fsck/k9/mail/filter/FixedLengthInputStream.java

package com.fsck.k9.mail.filter;

import java.io.IOException;
import java.io.InputStream;
import java.util.Locale;

/**
 * A filtering InputStream that stops allowing reads after the given length has been read. This
 * is used to allow a client to read directly from an underlying protocol stream without reading
 * past where the protocol handler intended the client to read.
 */
public class FixedLengthInputStream extends InputStream {
    private final InputStream in;
    private final int length;
    private int count = 0;

    public FixedLengthInputStream(InputStream in, int length) {
        this.in = in;
        this.length = length;
    }

    @Override
    public int available() throws IOException {
        return length - count;
    }

    @Override
    public int read() throws IOException {
        if (count >= length) {
            return -1;
        }

        int d = in.read();
        if (d != -1) {
            count++;
        }
        return d;
    }

    @Override
    public int read(byte[] b, int offset, int length) throws IOException {
        if (count >= this.length) {
            return -1;
        }

        int d = in.read(b, offset, Math.min(this.length - count, length));
        if (d != -1) {
            count += d;
        }
        return d;
    }

    @Override
    public int read(byte[] b) throws IOException {
        return read(b, 0, b.length);
    }

    @Override
    public long skip(long n) throws IOException {
        long d = in.skip(Math.min(n, available()));
        if (d > 0) {
            count += d;
        }
        return d;
    }

    @Override
    public String toString() {
        return String.format(Locale.US, "FixedLengthInputStream(in=%s, length=%d)", in.toString(), length);
    }

    public void skipRemaining() throws IOException {
        while (available() > 0) {
            skip(available());
        }
    }
}

File: mail/common/src/main/java/com/fsck/k9/mail/filter/FixedLengthInputStream.kt
package com.fsck.k9.mail.filter

import java.io.IOException
import java.io.InputStream
import java.util.Locale
import org.apache.commons.io.IOUtils

/**
 * A filtering InputStream that stops allowing reads after the given length has been read. This
 * is used to allow a client to read directly from an underlying protocol stream without reading
 * past where the protocol handler intended the client to read.
 */
class FixedLengthInputStream(
    private val inputStream: InputStream,
    private val length: Int,
) : InputStream() {
    private var numberOfBytesRead = 0

    // TODO: Call available() on underlying InputStream if remainingBytes() > 0
    @Throws(IOException::class)
    override fun available(): Int {
        return remainingBytes()
    }

    @Throws(IOException::class)
    override fun read(): Int {
        if (remainingBytes() == 0) {
            return -1
        }

        val byte = inputStream.read()
        if (byte != -1) {
            numberOfBytesRead++
        }

        return byte
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray, offset: Int, length: Int): Int {
        if (remainingBytes() == 0) {
            return -1
        }

        val byte = inputStream.read(b, offset, length.coerceAtMost(remainingBytes()))
        if (byte != -1) {
            numberOfBytesRead += byte
        }

        return byte
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray): Int {
        return read(b, 0, b.size)
    }

    @Throws(IOException::class)
    override fun skip(n: Long): Long {
        val numberOfSkippedBytes = inputStream.skip(n.coerceAtMost(remainingBytes().toLong()))
        if (numberOfSkippedBytes > 0) {
            numberOfBytesRead += numberOfSkippedBytes.toInt()
        }

        return numberOfSkippedBytes
    }

    @Throws(IOException::class)
    fun skipRemaining() {
        IOUtils.skipFully(this, remainingBytes().toLong())
    }

    private fun remainingBytes(): Int {
        return length - numberOfBytesRead
    }

    override fun toString(): String {
        return String.format(Locale.ROOT, "FixedLengthInputStream(in=%s, length=%d)", inputStream.toString(), length)
    }
}
